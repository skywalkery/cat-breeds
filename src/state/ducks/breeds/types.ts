import type { QueryFilter } from '@ducks/types';

export type Breed = {
  id: string;
  name: string;
  lowerName: string;
  intelligence: number;
  energyLevel: number;
  description?: string;
  imageUrl?: string;
};

export type ApiBreedRequest = {
  page?: number;
  limit?: number;
};

export type ApiBreed = {
  id: string;
  name: string;
  intelligence: number;
  energy_level: number;
  description?: string;
  image?: {
    url: string;
  };
};

export type ApiBreedResponse = ApiBreed[];

export enum Sort {
  NameAsc = 'NameAsc',
  NameDesc = 'NameDesc',
  IntelligenceAsc = 'IntelligenceAsc',
  IntelligenceDesc = 'IntelligenceDesc',
  EnergyAsc = 'EnergyAsc',
  EnergyDesc = 'EnergyDesc',
}

export type BreedFilters = QueryFilter;

export default {
  BREEDS_FETCH: 'BREEDS_FETCH',
  BREEDS_FETCH_SUCCESS: 'BREEDS_FETCH_SUCCESS',
  BREEDS_FETCH_ERROR: 'BREEDS_FETCH_ERROR',
} as const;
