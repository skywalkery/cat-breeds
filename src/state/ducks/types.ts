export interface QueryFilter {
  q?: string;
}

export interface DataSettings<SortType> {
  limit?: number;
  page?: number;
  sort?: SortType;
}

export enum LoadState {
  Init = 'init',
  Loading = 'loading',
  Processed = 'processed',
}
