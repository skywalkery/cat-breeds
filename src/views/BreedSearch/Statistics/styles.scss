.container {
  opacity: 1;
  transition: opacity 0.3s;
  max-width: 100%;

  @media (min-width: 48em) {
    max-width: 60%;
  }
}

.container-hidden {
  composes: container;
  opacity: 0;
}
