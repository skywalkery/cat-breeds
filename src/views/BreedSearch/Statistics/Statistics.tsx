import { FC, useMemo } from 'react';
import { useSelector } from 'react-redux';

import { breedsSelectors } from '@ducks/breeds';
import type { Breed, BreedFilters } from '@ducks/breeds/types';
import styles from './styles.scss';

type IntelligenceStatProps = {
  breeds: Breed[];
  type: 'Most' | 'Least';
};

const IntelligenceStat: FC<IntelligenceStatProps> = ({ breeds, type }) => {
  const names = useMemo(() => breeds.map((breed) => breed.name).join(', '), [
    breeds,
  ]);

  return (
    <div>
      {type} Intelligent cat breed{breeds.length > 1 ? 's' : ''}: {names}, with
      intelligence of {breeds[0]?.intelligence}
    </div>
  );
};

type Props = {
  filters: BreedFilters;
};

const Statistics: FC<Props> = ({ filters }) => {
  const mostIntelligence = useSelector((state) =>
    breedsSelectors.mostIntelligence(state, { filters })
  );
  const leastIntelligence = useSelector((state) =>
    breedsSelectors.leastIntelligence(state, { filters })
  );

  const isShown = !!mostIntelligence.length;

  return (
    <div className={styles[`container${isShown ? '' : '-hidden'}`]}>
      <IntelligenceStat type="Most" breeds={mostIntelligence} />
      <IntelligenceStat type="Least" breeds={leastIntelligence} />
    </div>
  );
};

export default Statistics;
