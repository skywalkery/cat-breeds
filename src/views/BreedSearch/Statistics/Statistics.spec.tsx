/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import { LoadState } from '@ducks/types';
import Statistics from './Statistics';

const breeds = {
  loadState: LoadState.Processed,
  data: [
    { id: '1', name: 'Aaa', lowerName: 'aaa', intelligence: 1, energyLevel: 2 },
    { id: '2', name: 'Bbb', lowerName: 'bbb', intelligence: 1, energyLevel: 2 },
    { id: '3', name: 'Ccc', lowerName: 'ccc', intelligence: 2, energyLevel: 1 },
    { id: '4', name: 'Ddd', lowerName: 'ddd', intelligence: 2, energyLevel: 1 },
  ],
};

describe('Breed Statistics', () => {
  it('renders properly without filter', () => {
    const store = createStore(() => ({
      breeds,
    }));

    const component = renderer.create(
      <Provider store={store}>
        <Statistics filters={{}} />
      </Provider>
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders properly with filter', () => {
    const store = createStore(() => ({
      breeds,
    }));

    const component = renderer.create(
      <Provider store={store}>
        <Statistics filters={{ q: 'A' }} />
      </Provider>
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
