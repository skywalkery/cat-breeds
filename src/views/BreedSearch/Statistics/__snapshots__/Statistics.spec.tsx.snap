// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breed Statistics renders properly with filter 1`] = `
<div
  className="container"
>
  <div>
    Most
     Intelligent cat breed
    
    : 
    Aaa
    , with intelligence of 
    1
  </div>
  <div>
    Least
     Intelligent cat breed
    
    : 
    Aaa
    , with intelligence of 
    1
  </div>
</div>
`;

exports[`Breed Statistics renders properly without filter 1`] = `
<div
  className="container"
>
  <div>
    Most
     Intelligent cat breed
    s
    : 
    Ccc, Ddd
    , with intelligence of 
    2
  </div>
  <div>
    Least
     Intelligent cat breed
    s
    : 
    Aaa, Bbb
    , with intelligence of 
    1
  </div>
</div>
`;
